### ISSUE TITLE
[Quantized Depth Completion] Questions about implementation details

### ISSUE THREAD

**Issue by james77777778** [sony/ai-research-code#61]
First of all, thanks for the great work, but the source code is still missing.
Could you share the training/evaluating code and pretrained weights about this work?

Also, I'm trying to reimplement with PyTorch and I have some questions about the paper:
1. How to compute the surface normal from ground truth depth in NYU Depth v2? The paper only shows the approximation for training but not the accurate one?!
2. The dot pattern to produce sparse depth in NYU Depth v2 is unknown. Can you share the example to reproduce?
3. The kernel_size of MaxPooling2D is missing
4. The kernel_size, filter_size of Conv2d in Upsampling layer is missing

Thanks! Look forward to your kind reply.

**Comment by VC86** [comment]
Dear HongYu,

thank you for your interest!

_- Q: Could you share the training/evaluating code and pretrained weights about this work?_
- A: Thanks for your interest again; as this code originates from a commercial research entity, the code contains non-shareable elements that would require a license (it is copyright Sony 2022). The intention of this landing page is to provide links/citation and share the evaluation set promised in the paper.

_- Q: How to compute the surface normal from ground truth depth in NYU Depth v2? The paper only shows the approximation for training but not the accurate one?!_
- A: The surface normals can be estimated by _optionally_ smoothing the NYU GT, and by applying the same normals estimation operator to both GT and data source, while accounting properly for invalid pixels (present in parts of NYU, notably in the extended dataset) and removing invalid normals from the normals loss computation. This is good enough for supervising the training process.

_- Q: The dot pattern to produce sparse depth in NYU Depth v2 is unknown. Can you share the example to reproduce?_
- A: To share the code generating the dot pattern we would have to sign a license agreement with you for strict non-commercial use of the code (_e.g._, if you are in a university or non-profit organization) as that code, too, is copyrighted. Feel free to reach out (to [me](mailto:valerio.cambareri@sony.com)) if your use would comply with this.

_- Q: The kernel_size of MaxPooling2D is missing_
- A: The kernel size for the MaxPools is always 2, to reduce the resolution by a factor 2 in each dimension. This is indicated in the paper.

_- Q: The kernel_size, filter_size of Conv2d in Upsampling layer is missing_
- A: As specified in the paper, the kernel size of Conv2D is always 3 (except the last MLP which uses kernel size 1) and stride is always 1 with “same” padding. The upsampling factor is also always 2. The number of filters in the upsample – convolve stage (decoder) mirrors the number of filters in the encoder at the same scale, as per the graph in Figure 2b.

**Comment by james77777778** [comment]
Hi @VC86,
Thank you for your fast and informative reply!

May I ask further for the `Normals Estimation Block`?
I try to build it like this:

```python
import numpy as np
import torch
import torch.nn as nn
import torchvision.transforms.functional as TF

arr = np.array(range(25))
tensor = torch.from_numpy(arr).to(torch.float).reshape(1, 1, 5, 5)
grad_x_layer = nn.Conv2d(1, 1, kernel_size=(1, 3), stride=1, padding=(0, 1), bias=False)
grad_y_layer = nn.Conv2d(1, 1, kernel_size=(3, 1), stride=1, padding=(1, 0), bias=False)
with torch.no_grad():
    grad_x_layer.weight = nn.Parameter(torch.tensor((-0.5, 0, 0.5)).reshape((1, 1, 1, 3)))
    grad_y_layer.weight = nn.Parameter(torch.tensor((-0.5, 0, 0.5)).reshape((1, 1, 3, 1)))
    grad_x = grad_x_layer(tensor)
    grad_y = grad_y_layer(tensor)
    minus_1 = -1 * torch.ones_like(tensor)
    normals = torch.cat((grad_x, grad_y, minus_1), dim=1)
    normals = normals / torch.linalg.norm(normals, dim=1, ord=2).unsqueeze(1)
print('input:\n', tensor)
print('grad_x:\n', grad_x)
print('grad_y:\n', grad_y)
print('normals:\n', normals)
```

and the output:

```bash
input:
 tensor([[[[ 0.,  1.,  2.,  3.,  4.],
          [ 5.,  6.,  7.,  8.,  9.],
          [10., 11., 12., 13., 14.],
          [15., 16., 17., 18., 19.],
          [20., 21., 22., 23., 24.]]]])
grad_x:
 tensor([[[[  0.5000,   1.0000,   1.0000,   1.0000,  -1.5000],
          [  3.0000,   1.0000,   1.0000,   1.0000,  -4.0000],
          [  5.5000,   1.0000,   1.0000,   1.0000,  -6.5000],
          [  8.0000,   1.0000,   1.0000,   1.0000,  -9.0000],
          [ 10.5000,   1.0000,   1.0000,   1.0000, -11.5000]]]])
grad_y:
 tensor([[[[ 2.5000,  3.0000,  3.5000,  4.0000,  4.5000],
          [ 5.0000,  5.0000,  5.0000,  5.0000,  5.0000],
          [ 5.0000,  5.0000,  5.0000,  5.0000,  5.0000],
          [ 5.0000,  5.0000,  5.0000,  5.0000,  5.0000],
          [-7.5000, -8.0000, -8.5000, -9.0000, -9.5000]]]])
normals:
 tensor([[[[ 0.1826,  0.3015,  0.2649,  0.2357, -0.3094],
          [ 0.5071,  0.1925,  0.1925,  0.1925, -0.6172],
          [ 0.7333,  0.1925,  0.1925,  0.1925, -0.7868],
          [ 0.8433,  0.1925,  0.1925,  0.1925, -0.8701],
          [ 0.8113,  0.1231,  0.1161,  0.1098, -0.7692]],

         [[ 0.9129,  0.9045,  0.9272,  0.9428,  0.9283],
          [ 0.8452,  0.9623,  0.9623,  0.9623,  0.7715],
          [ 0.6667,  0.9623,  0.9623,  0.9623,  0.6052],
          [ 0.5270,  0.9623,  0.9623,  0.9623,  0.4834],
          [-0.5795, -0.9847, -0.9864, -0.9879, -0.6355]],

         [[-0.3651, -0.3015, -0.2649, -0.2357, -0.2063],
          [-0.1690, -0.1925, -0.1925, -0.1925, -0.1543],
          [-0.1333, -0.1925, -0.1925, -0.1925, -0.1210],
          [-0.1054, -0.1925, -0.1925, -0.1925, -0.0967],
          [-0.0773, -0.1231, -0.1161, -0.1098, -0.0669]]]])
```

Is it correct?

Because after directly applying this implementation on GT depth in NYU Depth v2, the result is strange compared to the visualization in the paper.
For example (`data/nyudepthv2/val/official/00001.h5`):
- RGB
    ![rgb](https://user-images.githubusercontent.com/20734616/192667903-ad57ece8-9476-48a9-a5b7-014b651feb1b.png)
- Normals from estimation (with GT depth)
    ![normals](https://user-images.githubusercontent.com/20734616/192667917-a9b1779e-9298-47dc-94ab-a098a9771f9b.png)

The minimal and reproducible snippets:
```python
import h5py
import numpy as np
import torch
import torch.nn as nn
import torchvision.transforms.functional as TF
from PIL import Image

with h5py.File('data/nyudepthv2/val/official/00001.h5', 'r') as f:
    gt_depth = torch.from_numpy(np.array(f['depth'], dtype=np.float32)).unsqueeze(0).unsqueeze(0)  # (B, 1, H, W)
    rgb_img = Image.fromarray(np.transpose(f['rgb'], (1, 2, 0)))
with torch.no_grad():
    grad_x = grad_x_layer(gt_depth)
    grad_y = grad_y_layer(gt_depth)
    minus_1 = -1 * torch.ones_like(gt_depth)
    normals = torch.cat((grad_x, grad_y, minus_1), dim=1)
    normals = normals / torch.linalg.norm(normals, dim=1, ord=2).unsqueeze(1)
    normals = ((normals + 1) / 2 * 255).squeeze().to(torch.uint8)

normals = TF.to_pil_image(normals).save('normals.png')
rgb_img.save('rgb.png')
```

Thank you so much!

**Comment by VC86** [comment]
Your code looks correct and the test above is also numerically correct, but the normals aren't as I would expect them when visualized, indeed (although the way you convert them to UINT8 also looks correct).
Some notes:
- In some flavors you may want to concatenate `(-grad_x, -grad_y, torch.ones_like(gt_depth))` depending on what convention you follow for the normals (as noted in the paper -- the conventions must match between GT and estimate).
- I convert the depth to millimeters before computing the normals.

**Comment by james77777778** [comment]
Thanks again for your reply!

From the information about centered differences (dz/dy, dz/dx), I modified the code as following:

```python
import h5py
import numpy as np
import torch
import torch.nn.functional as F
import torchvision.transforms.functional as TF
from PIL import Image

with h5py.File('data/nyudepthv2/val/official/00001.h5', 'r') as f:
    gt_depth = torch.from_numpy(np.array(f['depth'], dtype=np.float32)).unsqueeze(0).unsqueeze(0)  # (B, 1, H, W)
    rgb_img = Image.fromarray(np.transpose(f['rgb'], (1, 2, 0)))

# resize & center crop (480, 640) -> (240, 320) -> (224, 304)
gt_depth, rgb_img = TF.resize(gt_depth, (240, 320)), TF.resize(rgb_img, (240, 320))
gt_depth, rgb_img = TF.center_crop(gt_depth, (224, 304)), TF.center_crop(rgb_img, (224, 304))

# take the scale into account
scaled_gt_depth = gt_depth / torch.amax(gt_depth) * 255.0

# compute normals
grad_x_weights = torch.tensor((-0.5, 0, 0.5), dtype=torch.float, requires_grad=False)
grad_x_weights = grad_x_weights.reshape((1, 1, 1, 3))
grad_y_weights = torch.tensor((-0.5, 0, 0.5), dtype=torch.float, requires_grad=False)
grad_y_weights = grad_y_weights.reshape((1, 1, 3, 1))
with torch.no_grad():
    x_padded_dense_depth = F.pad(scaled_gt_depth, (1, 1, 0, 0), 'replicate')
    y_padded_dense_depth = F.pad(scaled_gt_depth, (0, 0, 1, 1), 'replicate')
    grad_x = F.conv2d(x_padded_dense_depth, grad_x_weights)
    grad_y = F.conv2d(y_padded_dense_depth, grad_y_weights)
    minus_1 = -1 * torch.ones_like(scaled_gt_depth)
    normals = torch.cat((grad_x, grad_y, minus_1), dim=1)
    normals = normals / torch.linalg.norm(normals, dim=1, ord=2).unsqueeze(1)

# visualization
print(f'scaled_gt_depth stats: min={torch.min(scaled_gt_depth):.2f}, max={torch.max(scaled_gt_depth):.2f}')
print(f'normals stats        : min={torch.min(normals):.2f}, max={torch.max(normals):.2f}')
TF.to_pil_image(((normals + 1) / 2 * 255).squeeze().to(torch.uint8)).save('normals.png')
rgb_img.save('rgb.png')
```

and the output:
```bash
scaled_gt_depth stats: min=131.15, max=255.00
normals stats        : min=-1.00, max=1.00
```

![normals2](https://user-images.githubusercontent.com/20734616/192723540-cce2da65-1e04-4a91-a20a-c28379c51992.png)

I think the visualization is far better than previous one.
Is the procedure of scaling correct? (I'm not familiar with surface normals 😔)
`scaled_gt_depth = gt_depth / torch.amax(gt_depth) * 255.0`

Ref: [https://github.com/Ruthrash/surface_normal_filter](https://github.com/Ruthrash/surface_normal_filter)

**Comment by zherlock030** [comment]
@james77777778 Did u replicate the results of this paper? Is it good after quantization
